devito/arguments.py

```diff
@@ -0,0 +1,272 @@
+        """The data type of the iteration variable"""
+        return np.int32
+
+    @cached_property
+    def rtargs(self):
+        # TODO: Create proper Argument objects - with good init values
+        if self.size is not None:
+            return []
+        else:
+            size = ScalarArgument("%s_size" % self.name, self, max)
+            return [size]
+
+    @property
+    def ccode(self):
+        """C-level variable name of this dimension"""
+        return "%s_size" % self.name if self.size is None else "%d" % self.size
+
+    @property
+    def decl(self):
+        """Variable declaration for C-level kernel headers"""
+        return c.Value("const int", self.ccode)
+
+    # TODO: Do I need a verify on a dimension?
+    def verify(self, value):
+        verify = True
+
+        if value is None and self._value is not None:
+            return verify
+
+        if value is not None and value == self._value:
+            return verify
+
+        if self.size is not None:
+            # Assuming the only people calling my verify are symbolic data, they need to
+            # be bigger than my size if I have a hard-coded size
+            verify = (value >= self.size)
+        else:
+            # Assuming self._value was initialised as maxint
+            if value is not None and self._value is not None:
+                value = self.reducer(self._value, value)
+            if hasattr(self, 'parent'):
+                verify = verify and self.parent.verify(value)
+                # If I don't know my value, ask my parent
+                if value is None:
+                    value = self.parent.value
+
+            # Derived dimensions could be linked through constraints
+            # At this point, a constraint needs to be added that enforces
+            # dim_e - dim_s < SOME_MAX
+            # Also need a default constraint that dim_e > dim_s (or vice-versa)
+            verify = verify and all([a.verify(v) for a, v in zip(self.rtargs, (value,))])
+            if verify:
+                self._value = value
+        return verify
+
+
+class SymbolicDataArgProvider(RuntimeArgProvider):
+
+    """ Class used to decorate Symbolic Data objects with behaviour required for runtime
+        arguments.
+    """
+
+    @cached_property
+    def rtargs(self):
+        return [TensorArgument(self.name, self, self.dtype)]
+
+
+class ScalarFunctionArgProvider(RuntimeArgProvider):
+
+    """ Class used to decorate Scalar Function objects with behaviour required for runtime
+        arguments.
+    """
+
+    @cached_property
+    def rtargs(self):
+        return [ScalarArgument(self.name, self, self.dtype)]
+
+
+class TensorFunctionArgProvider(RuntimeArgProvider):
+
+    """ Class used to decorate Tensor Function objects with behaviour required for runtime
+        arguments.
+    """
+
+    @cached_property
+    def rtargs(self):
+        return [TensorArgument(self.name, self, self.dtype)]
+
+
+def log_args(arguments, log=debug):
```

Comment:
Be default it will write to the `debug` log level, but that can be over-riden with the parameter. What's wrong here?